### PR TITLE
Add a new keyword, "calldata", to allow explicitly specify data location in external function's argument list

### PR DESCRIPTION
**Pull request by sifmelcara**
As mentioned in #4173 
Maybe this should be merged into `0.5.0` branch?

### CODE REVIEW

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 1]
```diff
@@ -6225,7 +6225,7 @@ BOOST_AUTO_TEST_CASE(warn_unspecified_storage)
 			}
 		}
 	)";
-	CHECK_ERROR(text, TypeError, "Storage location must be specified as either \"memory\" or \"storage\".");
+	CHECK_ERROR(text, TypeError, "Data location must be specified as either \"memory\" or \"storage\".");
```
Comment (axic) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:6109]: These could be pulled out as non-breaking changes.

File: libsolidity/parsing/Parser.cpp [hunk 2]
```diff
@@ -592,11 +592,22 @@ ASTPointer<VariableDeclaration> Parser::parseVariableDeclaration(
...
+						location = VariableDeclaration::Location::Storage;
+						break;
+					case Token::Memory:
+						location = VariableDeclaration::Location::Memory;
+						break;
+					case Token::CallData:
+						location = VariableDeclaration::Location::CallData;
+						break;
+					default:
+						solAssert(false, "Unknown storage location.");
```
Comment (axic) [inline libsolidity/parsing/Parser.cpp]: I think this should read `Unknown data location.`

**Comment by axic** [comment]
The test cases (apart from `calldata`) cold also be merged straight away.

**Comment by sifmelcara** [comment]
It seems `050` have just been merged into `develop`. Do I still need to open another PR for the non-backward compatible change?

**Review by axic — CHANGES_REQUESTED** [review]
Since you introduce a new token, can you please include that in the grammar?

**Comment by axic** [comment]
I think we should be fine not splitting it. Can you please rebase it?

**Comment by sifmelcara** [comment]
Ok. `calldata` added to `docs/grammar.txt` (are there other files I should change?). Rebased against `develop`.

**Comment by axic** [comment]
Sorry, it also needs a changelog entry :)

**Comment by sifmelcara** [comment]
Hmm.. Looks like the tests run on circleci have some random failure? I only changed Changelog.md, which should not make it became failed.

File: test/libsolidity/syntaxTests/dataLocations/function_argument_location_specifier_test.sol [hunk 3]
```diff
@@ -0,0 +1,8 @@
+contract test {
+    function fa(bytes) public;
+    function fb(bytes memory) public;
+    function fc(bytes storage) public;
+    function fd(bytes calldata) public;
+}
+// ----
+// TypeError: (101-106): Location has to be memory for publicly visible functions (remove the "storage" or "calldata" keyword).
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/dataLocations/function_argument_location_specifier_test.sol]: Why is there only one error? There should be two, one for `fc`, one for `fd`.

Comment (chriseth) [reply test/libsolidity/syntaxTests/dataLocations/function_argument_location_specifier_test.sol]: Ah it is a fatal one! In that case, please change this test case to have exactly one function.

File: test/libsolidity/syntaxTests/dataLocations/function_argument_location_specifier_test_external.sol [hunk 4]
```diff
@@ -0,0 +1,8 @@
+contract test {
+    function fa(bytes) external;
+    function fb(bytes memory) external;
+    function fc(bytes storage) external;
+    function fd(bytes calldata) external;
+}
+// ----
+// TypeError: (65-70): Location has to be calldata for external functions (remove the "memory" or "storage" keyword).
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/dataLocations/function_argument_location_specifier_test_external.sol]: Same here.

**Review by chriseth — CHANGES_REQUESTED** [review]
Please split negative test cases such that they are as small as possible to produce the error.

**Comment by chriseth** [comment]
Please rebase since #4173  has been merged.

**Comment by sifmelcara** [comment]
Those are some good advice :)
Tests updated and rebased.

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
@axic fine to merge?

**Review by axic — CHANGES_REQUESTED** [review]
Sorry, haven't noticed, but please add a test in `SolidityTypes`:`type_identifiers`. There is an explicit test for memory and in implicit one for storage. Please add an explicit one for storage and calldata.

**Comment by sifmelcara** [comment]
Hmm... No idea why the test timed out.

**Review by chriseth — APPROVED** [review]

**Review by chriseth — APPROVED** [review]

**Pull request merged**